pyflakes/checker.py

```diff
@@ -734,8 +734,44 @@ def NAME(self, node):
             # arguments, but these aren't dispatched through here
             raise RuntimeError("Got impossible expression context: %r" % (node.ctx,))
 
+    def CONTINUE(self, node):
+        n = node
+        while hasattr(n, 'parent'):
+            n = n.parent
+            if isinstance(n, (ast.While, ast.For)):
+                # Doesn't apply unless it's in the loop itself
+                in_else = False
+                for else_node in n.orelse:
+                    for child in ast.walk(else_node):
+                        if child == node:
+                            in_else = True
+                            break
+                    if in_else:
+                        break
+                else:
+                    return
+            if isinstance(n, (ast.FunctionDef, ast.ClassDef)):
+                break
+            # Handle Try/TryFinally difference in Python < and >= 3.3
+            if hasattr(n, 'finalbody') and isinstance(node, ast.Continue):
+                in_finally = False
+                for finally_node in n.finalbody:
+                    for child in ast.walk(finally_node):
+                        if child == node:
+                            in_finally = True
+                            break
+                    if in_finally:
+                        self.report(messages.ContinueInFinally, node)
+                        return
+        if isinstance(node, ast.Continue):
+            self.report(messages.ContinueOutsideLoop, node)
+        else:  # ast.Break
+            self.report(messages.BreakOutsideLoop, node)
```

Comment:
That's roughly the algorithm (and more or less what I had in the first pass), but not completely correct. More accurate is "continue or break is only valid if it is in a loop, and if it a function or class definition, the loop must also be inside the function or class definition. And also for continue if it is in a finally block the loop must also be inside the finally block." I'm not sure if that can be written as a simple predicate (but I may just need to think about it harder). 

I guess the algorithm itself is not too complicated (it's just, walk up the tree and note what you see first, a loop, a function definition, a class definition, or (for `continue`) a finally block). The complications are arising from the fact that for loops and finally blocks, the nodes have different subtrees, so just saying "the `continue` is in the `ast.For` is not enough because it could be in the `orelse` part. So maybe that's what really should be abstracted here.  

Thoughts?